### PR TITLE
feat: try to clean up and simplify the file adding interface

### PR DESCRIPTION
**Pull request by djgilcrease**
The goal is to make this backward compatible with current config files so people can stay with the simplified format until they need more advanced features. 

Internally all packagers `ONLY` use the Contents data which is a merging of all the old file specifications into one list. IF a file was specified in a packager specific override (eg `rpm.Files`) then the `content.Packager` field is set to the correct packager automatically which ensures files for other packagers do not get mixed together.

Example
```yaml
name: "contents foo"
arch: "amd64"
version: "v1.2.3"
contents:
  - src: "./testdata/*.conf"
    dst: "/etc/foo/"
    type: "config|noreplace"
    file_info:
      mode: 0644
      mtime: 2008-01-02T15:04:05Z
      owner: me
      group: whatever
  - src: "./testdata/whatever2.conf"
    dst: "/etc/rpm/whatever.conf"
    type: "config|noreplace"
    packager: "rpm"
    file_info:
      mode: 0644
  - src: "./testdata/whatever2.conf"
    dst: "/etc/apk/whatever.conf"
    type: "config|noreplace"
    packager: "apk"
    file_info:
      mode: 0644
  - src: "./testdata/whatever2.conf"
    dst: "/etc/deb/whatever.conf"
    type: "config|noreplace"
    packager: "deb"
    file_info:
      mode: 0644
```

The `files`, `config_files`, `symlinks`, `rpm.config_noreplace_files`, and `rpm.ghost_files` are merged into `contents` and all packagers only the the data from `contents` now. This makes it simple for people to leave their config files as is and not break anything, but as they need more advanced features, like control over the file mode, owner, group, etc they can convert all their files to the contents as well.

I have also set it up so the simple mapping will work as well
```yaml
contents:
  a: b
  a2: b2
```

Also this is setup to be ready to handle contents as a string in prep for #43 but is not implemented
```yaml
contents: /path/to/a/tgz
```

### CODE REVIEW

**Comment by erikgeiser** [comment]
This is a pretty significant change and should definitely be discussed. I do think your approach is cleaner and more elegant but the old one is probably more pragmatic, since most people probably have a whole bunch of regular files and rarely any special files. I have no data to back that up though. I generate my configs automatically as a part of a bigger custom build process, but for those writing configs by hand, it will be much more to type.

I'll have a look at the exact changes when I find some time, maybe tomorrow.

**Comment by djgilcrease** [comment]
Ya, that's why I am making sure to ensure backwards compatibility so most people can continue to use the simple config format if they want.

**Comment by erikgeiser** [comment]
If I read your code correctly you planned to implement a custom unmarshal method so `files` can be both a YAML mapping and a list of `FileToCopy` in the YAML file. This is a great way to implement it.

Maybe we can use `src` and `dst` as YAML keys to make it more convenient for regular files in the new format.

Also I dislike the name of the type `FileToCopy`. If it's a glob, it represents multiple files that need to be copied. If it's a ghost file, nothing is to be copied. Maybe we could rename it to `Content`.

File: internal/files/files.go [hunk 1]
```diff
@@ -1,38 +1,78 @@
...
 			return nil, err
 		}
-		for src, dst := range globbed {
-			files = append(files, FileToCopy{src, dst})
+		appendAndSort(globbed, files)
+	}
+
+	return files, nil
+}
+// Expand gathers all of the real files to be copied into the package.
```
Comment (caarlos0) [inline internal/files/files.go:50]: same godocs as Expand..

File: internal/files/files.go [hunk 2]
```diff
@@ -1,38 +1,78 @@
...
-	Source      string
-	Destination string
+	Source string `yaml:"source,omitempty"`
+	Destination string `yaml:"destination,omitempty"`
+	Type string `yaml:"type,omitempty"`
+	Packager string `yaml:"packager,omitempty`
+	Mode uint `yaml:"mode,omitempty"`
 }
 
+//type tmpFruitBasket []map[string]yaml.Node
```
Comment (caarlos0) [inline internal/files/files.go:20]: rm commented code

**Comment by erikgeiser** [comment]
This is also nice for #43 because now we can then completely decide what do based on the YAML node type:
* **Mapping:** `src`->`dst` mapping of regular files, no non-regular files supported (only via depecated options), this is nice for very simple packages
* **Sequence:** The new rich file types you introduced here
* **String:** Path to a `.tar.gz` file

I think then `Content` or `Contents` would indeed be a much more fitting name.

File: internal/files/files.go (hunk 1 above)
Comment (djgilcrease) [reply internal/files/files.go]: removed the expand function

File: internal/files/files.go (hunk 2 above)
Comment (djgilcrease) [reply internal/files/files.go]: done

**Comment by djgilcrease** [comment]
If we want a new field instead of trying to override the current `files` field that should actually be easier and cleaner than what I am having to do now I think. I can still have code that merges the `files`, `config_files`, `symlinks`, `rpm.config_noreplace_files`, and `rpm.ghost_files` into `contents` so all the packagers only use the `contents`

```yaml
contents:
  - src: "something"
    dst: "/opt/pkg/something"
    type: "config"
    packager: "rpm"
    mode: 0644
```

give me a day or so to play with that concept to see if it turns out cleaner

**Comment by djgilcrease** [comment]
> give me a day or so to play with that concept to see if it turns out cleaner

Much cleaner code now that I am not trying to mess with existing fields

File: www/docs/configuration.md [hunk 3]
```diff
@@ -85,31 +85,75 @@ suggests:
 # Packages it conflicts with. (overridable)
 conflicts:
   - mercurial
-
-# Files to add to the package. (overridable)
+    
+# Contents to att to the package
```
Comment (caarlos0) [inline www/docs/configuration.md:89]:
```suggestion
# Contents to add to the package
```

**Review by caarlos0 — APPROVED** [review]
this looks awesome, thanks for working on this @djgilcrease ! 🚀

**Comment by caarlos0** [comment]
> The goal is to make this backward compatible with current config files so people can stay with the simplified format until they need more advanced features.

I would say we deprecate the old format. The new format is simple enough for the basic use case as well, and then when users need to change it, its not that much of a change?

wdyt?

if you agree, we can copy the deprecate package from goreleaser and do basically the same thing here, and remove the old option after 6mo or so. (and we can make this all in another PR since this one is already pretty big)

**Pull request merged**